Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,851 @@
+    ///
+    /// - Note: Calling this function more than once has no effect.
+    /// - Parameter failure: The failure indicating why the writer finished.
+    @inlinable
+    public func finish(with failure: Failure) {
+        self._storage.finish(with: failure)
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    /// The underlying sink of the ``NIOAsyncWriter``. This type allows to set the writability of the ``NIOAsyncWriter``.
+    ///
+    /// - Important: Once all copies to the ``NIOAsyncWriter/Sink`` are destroyed the ``NIOAsyncWriter`` will get finished.
+    public struct Sink {
+        /// This class is needed to hook the deinit to observe once all references to the ``NIOAsyncWriter/Sink`` are dropped.
+        @usableFromInline
+        /* fileprivate */ internal final class InternalClass: Sendable {
+            @usableFromInline
+            /* fileprivate */ internal let _storage: Storage
+
+            @inlinable
+            init(storage: Storage) {
+                self._storage = storage
+            }
+
+            @inlinable
+            deinit {
+                // We need to call finish here to resume any suspended continuation.
+                self._storage.finish(with: nil)
+            }
+        }
+
+        @usableFromInline
+        /* private */ internal let _internalClass: InternalClass
+
+        @usableFromInline
+        /* private */ internal var _storage: Storage {
+            self._internalClass._storage
+        }
+
+        @inlinable
+        init(storage: Storage) {
+            self._internalClass = .init(storage: storage)
+        }
+
+        /// Sets the writability of the ``NIOAsyncWriter``.
+        ///
+        /// If the writer becomes writable again all suspended yields will be resumed and the produced elements will be forwarded via
+        /// the ``NIOAsyncWriterDelegate/didYield(contentsOf:)`` method. If the writer becomes unwritable all
+        /// subsequent calls to ``NIOAsyncWriterDelegate/didYield(contentsOf:)`` will suspend.
+        ///
+        /// - Parameter writability: The new writability of the ``NIOAsyncWriter``.
+        public func setWritability(to writability: Bool) {
+            self._storage.setWritability(to: writability)
+        }
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    /// This is the underlying storage of the writer. The goal of this is to synchronize the access to all state.
+    @usableFromInline
+    /* fileprivate */ internal final class Storage: @unchecked Sendable {
+        /// The lock that protects our state.
+        @usableFromInline
+        /* private */ internal let _lock = Lock()
+        /// The counter used to assign an ID to all our yields.
+        @usableFromInline
+        /* private */ internal let _yieldIDCounter = ManagedAtomic<UInt64>(0)
```

Comment:
Can we wrap this in a `struct` that gives us back an object that isn't a `UInt64` that has the API surface we want? The way we use this made me think for a while that it was supposed to be a monotonically increasing value, but in actuality it just needs to be a _unique_ value. Having a type that encapsulates this logic and doesn't encourage comparison would be helpful.